### PR TITLE
452,446,447 vde master

### PR DESCRIPTION
**Pull request by rpoletaev**
3 tasks are combined, because in each subsequent used code the previous

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: packages/daemons/common.go [hunk 1]
```diff
@@ -165,3 +165,14 @@ func getHostPort(h string) string {
 	}
 	return fmt.Sprintf("%s:%d", h, consts.DEFAULT_TCP_PORT)
 }
+
+func getDaemonsToStart() []string {
+	if *conf.IsVDEMasterMode || *conf.IsVDEMode {
```
Comment (potehinre) [inline packages/daemons/common.go]: Mabye it is better to do some method in config like IsVDEMode() { return *conf.IsVDEMasterMode || conf.IsVDEMode }

File: packages/api/route.go [hunk 2]
```diff
@@ -86,6 +80,18 @@ func Route(route *hr.Router) {
 	post(`test/:name`, ``, getTest)
 	post(`content`, `template:string`, jsonContent)
 
+	if !*conf.IsVDEMasterMode && !*conf.IsVDEMode {
```
Comment (potehinre) [inline packages/api/route.go]: if !conf.IsVDEMode()

File: packages/api/api.go [hunk 3]
```diff
@@ -226,9 +226,9 @@ func DefaultHandler(method, pattern string, params map[string]int, handlers ...a
 		for _, par := range ps {
 			data.params[par.Key] = par.Value
 		}
-		vde := r.FormValue(`vde`)
-		if vde == `1` || vde == `true` {
-			data.vm = smart.GetVM(true, data.ecosystemId)
+
+		if *conf.IsVDEMasterMode || *conf.IsVDEMode {
```
Comment (potehinre) [inline packages/api/api.go]: if *conf.IsVDEMode

File: packages/conf/conf.go [hunk 4]
```diff
@@ -135,6 +148,28 @@ func SaveConfig() error {
...
+	if os.IsNotExist(err) {
+		cf, err = os.Create(path)
+		if err != nil {
+			log.WithFields(log.Fields{"type": consts.IOError, "error": err}).Error("Create config file failed")
+			return err
+		}
+	} else {
+		cf, err = os.Open(path)
+		if err != nil {
+			log.WithFields(log.Fields{"type": consts.IOError, "error": err}).Error("Create config file failed")
```
Comment (potehinre) [inline packages/conf/conf.go]: Opening config file failed

File: packages/daylight/daemonsctl/daemonsctl.go [hunk 5]
```diff
@@ -28,6 +28,10 @@ func RunAllDaemons() error {
 	log.Info("start daemons")
 	daemons.StartDaemons()
 
+	if *conf.IsVDEMasterMode || *conf.IsVDEMode {
```
Comment (potehinre) [inline packages/daylight/daemonsctl/daemonsctl.go]: conf.IsVDEMode()

File: packages/daylight/start.go [hunk 6]
```diff
@@ -252,6 +253,19 @@ func Start() {
 			log.WithFields(log.Fields{"type": consts.DBError, "error": err}).Error("InitDB")
 			Exit(1)
 		}
+
+		if *conf.IsVDEMasterMode || *conf.IsVDEMode {
```
Comment (potehinre) [inline packages/daylight/start.go]: conf.IsVDEMode()

File: packages/daylight/start.go [hunk 7]
```diff
@@ -252,6 +253,19 @@ func Start() {
...
+			err := model.ExecSchemaEcosystem(nil, 1, conf.Config.KeyID, ``, conf.Config.KeyID)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.DBError, "error": err}).Error("executing ecosystem schema")
+				Exit(1)
+			}
+
+			if err := model.CreateVDEIfNotExists(consts.DefaultVDE, conf.Config.KeyID); err != nil {
+				log.WithFields(log.Fields{"type": consts.DBError, "error": err}).Error("")
```
Comment (potehinre) [inline packages/daylight/start.go]: Error("Creating VDE If Not Exists")

File: packages/model/db.go [hunk 8]
```diff
@@ -160,6 +160,32 @@ func ExecSchemaLocalData(id int, wallet int64) error {
...
+
+	sp := &StateParameter{}
+	sp.SetTablePrefix(converter.Int64ToStr(id))
+	if _, err := sp.Get(nil, `founder_account`); err != nil {
+		log.WithFields(log.Fields{"type": consts.DBError, "error": err}).Error("creating vde")
+		return err
+	}
+
+	if converter.StrToInt64(sp.Value) != wallet {
+		return fmt.Errorf("access denied")
```
Comment (potehinre) [inline packages/model/db.go]: logging?

File: packages/script/vm.go [hunk 9]
```diff
@@ -814,11 +815,13 @@ func (rt *RunTime) RunCode(block *Block) (status int, err error) {
 func (rt *RunTime) Run(block *Block, params []interface{}, extend *map[string]interface{}) (ret []interface{}, err error) {
 	defer func() {
 		if r := recover(); r != nil {
+			log.Error(r)
 			rt.vm.logger.WithFields(log.Fields{"type": consts.PanicRecoveredError, "stack": string(debug.Stack())}).Error("runtime panic error")
 			err = fmt.Errorf(`runtime panic error`)
 		}
 	}()
 	info := block.Info.(*FuncInfo)
+	log.Infof("INFOF: %+v", *info)
```
Comment (potehinre) [inline packages/script/vm.go]: Debug logging is prohibited in PRs

File: packages/script/vm.go [hunk 10]
```diff
@@ -814,11 +815,13 @@ func (rt *RunTime) RunCode(block *Block) (status int, err error) {
 func (rt *RunTime) Run(block *Block, params []interface{}, extend *map[string]interface{}) (ret []interface{}, err error) {
 	defer func() {
 		if r := recover(); r != nil {
+			log.Error(r)
```
Comment (potehinre) [inline packages/script/vm.go]: Debug logging is prohibited

File: packages/smart/funcs.go [hunk 11]
```diff
@@ -1187,3 +1203,56 @@ func UpdateCron(sc *SmartContract, id int64) error {
 
 	return nil
 }
+
+// CreateVDE allow create new VDE throw vdemanager
+func CreateVDE(sc *SmartContract, name, dbUser, dbPassword string, port int64) error {
+
+	log.Infoln("FROM CREATEVDE FUNC")
```
Comment (potehinre) [inline packages/smart/funcs.go]: Debug logging is prohibited

File: packages/smart/funcs.go [hunk 12]
```diff
@@ -1187,3 +1203,56 @@ func UpdateCron(sc *SmartContract, id int64) error {
...
+		log.WithFields(log.Fields{"type": consts.WrongModeError, "error": err}).Error("create new VDE")
+		return err
+	}
+
+	return vdemanager.Manager.DeleteVDE(name)
+}
+
+// StartVDE run VDE process
+func StartVDE(sc *SmartContract, name string) error {
+	if vdemanager.Manager == nil {
```
Comment (potehinre) [inline packages/smart/funcs.go]: code duplication - move it to separate function

File: packages/smart/funcs.go [hunk 13]
```diff
@@ -1187,3 +1203,56 @@ func UpdateCron(sc *SmartContract, id int64) error {
...
+		log.WithFields(log.Fields{"type": consts.WrongModeError, "error": err}).Error("create new VDE")
+		return err
+	}
+
+	return vdemanager.Manager.StartVDE(name)
+}
+
+// StopVDE stops VDE process
+func StopVDE(sc *SmartContract, name string) error {
+	if vdemanager.Manager == nil {
```
Comment (potehinre) [inline packages/smart/funcs.go]: code duplication - move it to separate to separate function

File: packages/smart/funcs.go [hunk 14]
```diff
@@ -1187,3 +1203,56 @@ func UpdateCron(sc *SmartContract, id int64) error {
...
+		log.WithFields(log.Fields{"type": consts.WrongModeError, "error": err}).Error("create new VDE")
+		return err
+	}
+
+	return vdemanager.Manager.StopVDE(name)
+}
+
+// GetVDEList returns list VDE process with statuses
+func GetVDEList(sc *SmartContract, name string) (map[string]string, error) {
+	if vdemanager.Manager == nil {
```
Comment (potehinre) [inline packages/smart/funcs.go]: same stuff

File: packages/smart/funcs.go [hunk 15]
```diff
@@ -1187,3 +1203,56 @@ func UpdateCron(sc *SmartContract, id int64) error {
...
+		err := fmt.Errorf("node must be running as VDEMaster")
+		log.WithFields(log.Fields{"type": consts.WrongModeError, "error": err}).Error("create new VDE")
+		return nil, err
+	}
+
+	processes := vdemanager.Manager.ListProcess()
+	log.Infof("%+v", processes)
```
Comment (potehinre) [inline packages/smart/funcs.go]: debug logging is prohibited in PRs

File: packages/vdemanager/manager.go [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+
+	confEntry := pConf.NewConfigEntry(vdeDir)
+	confEntry.Name = "program:" + name
+	command := fmt.Sprintf("%s -VDEMode=true -initDatabase=true -configPath=%s -workDir=%s", bin(), vdeConfigPath, vdeDir)
+	confEntry.AddKeyValue("command", command)
+	proc := process.NewProcess("vdeMaster", confEntry)
+
+	mgr.processes.Add(name, proc)
+	mgr.processes.Find(name).Start(true)
+	log.Infoln(command)
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: Debug logging is prohibited

File: packages/vdemanager/manager.go [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+		log.WithFields(log.Fields{"type": consts.VDEManagerError, "error": err}).Error("on find VDE process")
+		return err
+	}
+
+	state := proc.GetState()
+	if state == process.STOPPED ||
+		state == process.EXITED ||
+		state == process.FATAL {
+		proc.Start(true)
+		log.Infof("VDE '%s' is started", name)
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: log.WithFields(log.Fields{"vde_name": name}).Info("VDE started")

File: packages/vdemanager/manager.go [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+		err := fmt.Errorf(`VDE '%s' is not exists`, name)
+		log.WithFields(log.Fields{"type": consts.VDEManagerError, "error": err}).Error("on find VDE process")
+		return err
+	}
+
+	state := proc.GetState()
+	if state == process.RUNNING ||
+		state == process.STARTING {
+		proc.Stop(true)
+		log.Infof("VDE '%s' is stoped", name)
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: log.WithFields(log.Fields{"vde_name": name})

File: packages/vdemanager/manager.go [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.IOError, "error": err, "path": childConfigsPath}).Error("Initialising VDE list")
+		return err
+	}
+
+	for _, item := range list {
+		if item.IsDir() {
+			procDir := path.Join(childConfigsPath, item.Name())
+			commandStr := fmt.Sprintf(commandTemplate, bin(), filepath.Join(procDir, consts.DefaultConfigFile), procDir)
+			log.Errorln("commandStr: ", commandStr)
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: Debug logging is prohibited in PRs

File: packages/vdemanager/manager.go [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+
+			proc := process.NewProcess("vdeMaster", confEntry)
+			Manager.processes.Add(item.Name(), proc)
+		}
+	}
+
+	return nil
+}
+
+func createKeyPair(privFilename, pubFilename string) (priv, pub []byte, err error) {
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: code duplicated from packages/install/first_block.go, maybe move to utils?

File: packages/vdemanager/manager.go [hunk 21]
```diff
@@ -0,0 +1,285 @@
...
+
+	err = createFile(pubFilename, []byte(hex.EncodeToString(pub)))
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func createFile(filename string, data []byte) error {
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: same stuff

File: packages/vdemanager/manager.go [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+	err := ioutil.WriteFile(filename, data, 0644)
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.IOError, "error": err}).Error("writing file")
+		return err
+	}
+	return nil
+}
+
+func bin() string {
+	return path.Join(conf.Config.WorkDir, "go-genesis")
```
Comment (potehinre) [inline packages/vdemanager/manager.go]: "go-genesis" must be constant

File: packages/vdemanager/manager_test.go [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+	if err := model.GormInit(conf.Config.DB.Host, conf.Config.DB.Port, conf.Config.DB.User, conf.Config.DB.Password, conf.Config.DB.Name); err != nil {
+		return err
+	}
+
+	return InitVDEManager()
+}
+func TestCreateVDE(t *testing.T) {
+
+	if err := prepare("/home/losaped/go/src/github.com/GenesisKernel/go-genesis/config.toml"); err != nil {
```
Comment (potehinre) [inline packages/vdemanager/manager_test.go]: /home/losaped/ in tests?

File: packages/vdemanager/manager_test.go [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+	if err := Manager.CreateVDE("one9", "one9", "one9", 8006); err != nil {
+		t.Error(err)
+		return
+	}
+
+}
+
+func TestListVDE(t *testing.T) {
+
+	if err := prepare("/home/losaped/go/src/github.com/GenesisKernel/go-genesis/config.toml"); err != nil {
```
Comment (potehinre) [inline packages/vdemanager/manager_test.go]: same stuff

File: packages/vdemanager/manager_test.go [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+		return
+	}
+
+	list := Manager.ListProcess()
+	fmt.Printf("%v\n", list)
+}
+
+func TestListRunningVDE(t *testing.T) {
+
+	if err := prepare("/home/losaped/go/src/github.com/GenesisKernel/go-genesis/config.toml"); err != nil {
```
Comment (potehinre) [inline packages/vdemanager/manager_test.go]: same stuff

File: packages/vdemanager/manager_test.go [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+		t.Error(err)
+		return
+	}
+
+	list := Manager.ListProcess()
+	fmt.Printf("%v\n", list)
```
Comment (potehinre) [inline packages/vdemanager/manager_test.go]: no prints in code

File: packages/vdemanager/manager_test.go [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+	if err := Manager.StartVDE("one3"); err != nil {
+		t.Error(err)
+		return
+	}
+
+	list = Manager.ListProcess()
+	fmt.Printf("%v\n", list)
```
Comment (potehinre) [inline packages/vdemanager/manager_test.go]: no prints in code

**Review by potehinre — CHANGES_REQUESTED** [review]

**Review by potehinre — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request closed without merging**